### ISSUE TITLE
[FeatReq] Add a Tasklist to Aider

### ISSUE THREAD

**Issue by Emasoft** [paul-gauthier/aider#959]
### Issue

Aider need a tasklist. This will help the progress and keeping track of with the code changes we need.
The best would be taking the example from Github: we need just three things:  
- feature requests (aka TODO)
- bug reports
- tasklist

The user should be able to create those with two new aider commands:  
  
**“/task feature <description>”** : add a feature request description
**“/task bug <description>”** : add a bug report
  
Also the AI should be able to create those entries by itself in the /ask mode if the user asks for new features, or if the user report a bug.

Those two kind of entries could also be imported from the github repo issues list, with a command like this:  
  
**“/github-sync <repo url>”**  

You should also be able to specify if you want just to sync the bugs and/or the features:  
  
**“/github-sync <repo url> issues bugs”**

When you need to have a overview of the project progress, you should be able to call a command like:  
  
**“/task list”**

...and getting a result like:  
  
**PROJECT TASKLIST:**

| n | Description                                            | Reference URL                                      | Status         |
|---------|--------------------------------------------------------|---------------------------------------------------|----------------|
| 1 | [❇️ Feat] Add user authentication feature                        | https://github.com/myproject/issues/45             | 📝 planned     |
| 2 | [🪲Bug] Fix crash on startup when configuration is missing     | https://github.com/myproject/issues/32             | ✅ solved      |
| 3 | [❇️ Feat] Implement dark mode for the application                |                                                   | 📝 planned     |
| 4 | [🪲Bug] Resolve memory leak in data processing module          | https://github.com/myproject/issues/19             | 🔄 in progress |
| 5 | [❇️ Feat] Add support for exporting data in CSV format           | https://github.com/myproject/issues/52             | 📝 planned     |
| 6 | [🪲Bug] Correct typo in user settings description              |                                                   | ✅ solved      |
| 7 | [❇️ Feat] Create an admin dashboard for monitoring activities    |                                                   | 📝 planned     |
| 8 | [🪲Bug] Address layout issues on mobile devices                | https://github.com/myproject/issues/27             |  📝 planned |
| 9 | [❇️ Feat] Integrate third-party API for weather updates          |                                                   | 📝 planned     |
| 10 | [🪲Bug] Fix broken link in the help documentation              |                                                   | ❌ wontfix     |

**Project Status Summary:**
- 📝 Planned: 6
- ✅ Solved: 2
- ❌ Wontfix: 1
- 🔄 In Progress: 1

**Write a task number and the new status (Example: “n solved”). “n drop” to delete a task. C to clear all tasks:**  
**> 8 solved**   

In Aider help:
*You can show the tasklist with the command* **/task list**
*You can delete a task with the command* **/task drop _n_**
*You can change the status of a task with the command* **/task status _n_ _new-status_**
*You can create new bug report with the command* **/task bug _description_**
*You can create new feature request with the command* **/task feature _description_**  
*You can sync bug reports and feature requests with your Github repo with*   **/github-sync**
  

NOTE: The AI can create tasks, but only the user can change the status of a task. Until a task is not marked as solved the AI will try to complete it, starting with fixing bugs, and only when all bugs are fixed it will start adding new features. Of course the user can ask the AI to focus on a certain task number.

What do you think?

**Comment by sholub89** [comment]
@Emasoft , wanted to mention - the original idea of `/ask` was an actual non-modify mode. So the tasklist edit has to be done somehow else. 

I wish there was a tasklist implementation, that would be super helpful. 
FYI there is a workaround I found on the web where people create a webpage with the tasks and add it to the context of Aider. Since it is a webpage - Aider can't modify it and it stays intact while you can refer to it saying work on task 1, fix the issue 6 etc.. But that isn't very convenient.

**Comment by paul-gauthier** [comment]
Thanks for trying aider and filing this issue.

Why not just have aider use a file like `TASKS.md` to create the task list? Then ask it work on items and mark them as done, etc?

**Comment by Emasoft** [comment]
No, for it to work it is necessary to implement a function calling of some sort. The AI should be able to select and retrieve the description of each task, and read its status.

**Comment by adamjosephjensen** [comment]
@Emasoft what would this prescriptive data model enable you to do that a TASKS.md (with links to external resources as needed, and with long descriptions also as-needed) not be able to do?

I am "just a guy" but it sounds like you are trying to front-load a lot of "shaping" / specifying work and then have an opinionated model that makes it easier to switch between work items.

I hope this isn't overly dramatic, but this seems like if the aider team built this they would be passing you a loaded shotgun pointed at your own foot. If your task descriptions are too big for one TASKS.md it might be better to look at minimizing work-in-progress instead of maximizing work-in-progress. Consider: by the time one work item is complete, the tasks for the next one need to change and the old ones are stale.

I have had a lot of success getting aider to write its own TASKS.md with all the detail it needs, and then work off of a TASKS.md file.
